Use userAgent instead of depricated appVersion to check for safari indexDB bug.

KoryNunn opened the pull request:
Currently firestore persistance is broken in cordova, as WKWebView's appVersion does not match Version/1[45], and appVersion is a depricated feature. This PR swaps to checking the userAgent, which is something cordova apps can affect, allowing some level of mitigation of this issue.

Fixes #6509

KoryNunn commented:
@dconeybe I have done the CLA, is there a way to re-run these jobs?

dconeybe commented:
> @dconeybe I have done the CLA, is there a way to re-run these jobs?

I'm not sure. Tagging @cherylEnkidu and @wu-hui, who were working with you on the related bug https://github.com/firebase/firebase-js-sdk/issues/6739

dconeybe commented:
Hi @KoryNunn. I think you may have linked to the wrong issue number in the description of this PR. You linked to https://github.com/firebase/firebase-js-sdk/issues/6739, which is why I tagged my team mates, but that issue doesn't appear to be related to appVersion or user agent at all. I think you meant to link to https://github.com/firebase/firebase-js-sdk/issues/6509. Is my thinking correct?

dconeybe commented:
Please merge in HEAD of the master branch and push it up. The GitHub actions have likely changed quite a bit.

KoryNunn commented:
> 

I did mean https://github.com/firebase/firebase-js-sdk/issues/6509 sorry.

I'll merge head shortly

dconeybe commented:
Thanks for updating your PR. Since this dark corner of the Firebase code base is unfamiliar with me, I've reached out to some colleagues for advice.

What do you think about leaving the original "if" condition in place, but adding an "or" clause? Basically, check appVersion _and_ userAgent and applying the workaround if at least one of the conditions is true? That would more likely preserve backwards compatibility IMO.

dconeybe commented on packages/firestore/src/local/indexeddb_persistence.ts:
```diff
@@ -977,7 +977,7 @@ export class IndexedDbPersistence implements Persistence {
         this.markClientZombied();
 
-        if (isSafari() && navigator.appVersion.match(/Version\/1[45]/)) {
+        if (isSafari() && navigator.userAgent.match(/Version\/1[45]/)) {
```
Let's do an "or" to check both the appVersion and userAgent.

i.e.

```ts
if (isSafari() && (navigator.appVersion.match(/Version\/1[45]/) || navigator.userAgent.match(/Version\/1[45]/)))
```

dconeybe replied:
@KoryNunn If you make this small change to the PR I'd be happy to merge it.

dconeybe reviewed (changes requested):
Please add a changeset using `yarn changeset`. It should be a patch release of `@firebase/firestore` and `firebase`.

For the changeset description, use something like this:

> Check `navigator.userAgent`, in addition to `navigator.appVersion`, when determining whether to work around an IndexedDb bug in Safari. Notably, this may be helpful for Cordova apps that lack `appVersion` in their webview yet need to apply the workaround.

KoryNunn commented:
I've updated this PR to more robustly work in both Safari and in WKWebView (the web view used in native apps).

In Safari, the match is Version/{number}, in WKWebView, the match is Mobile/{number}.

Also, this bug is present in version 16, so I added that to the check.

dconeybe commented on .changeset/chilled-buckets-sneeze.md:
```diff
@@ -0,0 +1,6 @@
+'firebase': patch
+---
+
+Check navigator.userAgent, in addition to navigator.appVersion, when determining whether to work around an IndexedDb bug in Safari. Add check for Mobile/# in addition to Version/# to include WKWebView safari versions in this check.
```
Please remove the 2nd sentence from the changeset. The first sentence is the right amount of information for the release notes. If they want more information they can click on the link from the release notes and see your full change.

dconeybe commented on packages/firestore/src/local/indexeddb_persistence.ts:
```diff
@@ -977,8 +977,8 @@ export class IndexedDbPersistence implements Persistence {
 
-        if (isSafari() && navigator.appVersion.match(/Version\/1[45]/)) {
-          // On Safari 14 and 15, we do not run any cleanup actions as it might
+        if (isSafari() && (navigator.appVersion.match(/(?:Version|Mobile)\/1[456]/) || navigator.userAgent.match(/(?:Version|Mobile)\/1[456]/))) {
```
Please factor out the regular expression into a local variable, to avoid duplicating it.

e.g. `const versionRegex = /(?:Version|Mobile)\/1[456]/;`

dconeybe replied:
Please run `yarn prettier` in the `packages/firestore` directory to format the code. It will do something wonky to try to get line wrapping at 80 characters.

dconeybe reviewed (commented):
LGTM! Will give approval once GitHub Actions pass. Thanks for your time (and a great deal of patience) in making this contribution to the SDK!

dconeybe reviewed (approved):
Please merge at your convenience.

KoryNunn commented:
"Only those with to this repository can merge pull requests."
